controlplane/kubeadm/internal/machine_filters.go

```diff
@@ -93,10 +94,34 @@ func OwnedControlPlaneMachines(controlPlaneName string) MachineFilter {
 		if controllerRef == nil {
 			return false
 		}
-		return controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == controlPlaneName
+		return controllerRef.Kind == "KubeadmControlPlane" && controllerRef.Name == owner.GetName() && controllerRef.UID == owner.GetUID()
 	}
 }
 
+func ControlPlaneMachines(clusterName string) func(machine *clusterv1.Machine) bool {
+	selector := ControlPlaneSelectorForCluster(clusterName)
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		return selector.Matches(labels.Set(machine.Labels))
+	}
+}
+
+// AdoptableControlPlaneMachines returns a MachineFilter function to find all un-controlled control plane machines.
+// Usage: managementCluster.GetMachinesForCluster(ctx, cluster, AdoptableControlPlaneMachines(cluster.Name, controlPlane))
+func AdoptableControlPlaneMachines(clusterName string) func(machine *clusterv1.Machine) bool {
+	return And(
+		func(machine *clusterv1.Machine) bool {
```

Comment:
I'm wondering if it would make sense to add new filters to ease the use of this:
```go
func IsNil(machine *clusterv1.Machine) bool {
	return machine == nil
}

func HasControllerRef(machine *clusterv1.Machine) bool {
	return metav1.GetControllerOf(machine) != nil
}
```

Then the test here could be:
```go
return And(
	Not(IsNil),
	ControlPlaneMachines(clusterName),
	Not(HasControllerRef),
)
```

Reply:
I can do this. At some point we're fighting the language a bit and causing a fair amount of allocation / pointer hopping, but since we're going to be filtering collections of maybe hundreds of things it's not going to matter :)